tensorflow/core/kernels/dml_reduce_ops.cc

```diff
@@ -66,19 +66,29 @@ class ReduceInitializationHelper : public InitializationHelper {
     return false;
   }
 
-  bool IsInputForwardable(OpKernelContext* ctx,
-                          absl::Span<const TensorShape> output_shapes,
-                          int& inputIndex) const override {
+  bool IsOutputForwardable(OpKernelContext* ctx,
```

Comment:
I think a more modern way to handle this would be to return an `absl::optional<int>` instead, that way you don't need to return a boolean AND also set an integer. This is also less error-prone (implementations can forget to set inputIndexToForward) and the caller doesn't need to check for `inputIndexToForward != -1`. The function could be called `GetForwardableInputIndex` or something similar.